### PR TITLE
Switched from etcd2 to etcd3

### LINKED ISSUE #245: Switch to etcd3
Hey, how about using etcd3 instead of the old 2?

### PR DESCRIPTION
**Pull request by saschagrunert**
Hey, this pull request not ready for merge. The main problem I got not solved right now is that all etcd3 clients needs to be connected to successfully start the etcd cluster. This means kubernetes will just come up when the etcd cluster is ready, which is currently not the provisioning approach you implemented. Maybe can you help me a bit with the Vagrantfile to first setup the etcd cluster, then deploy Kubernetes including the health checks?

Closes #245

### CODE REVIEW

File: Vagrantfile [hunk 1]
```diff
@@ -185,14 +185,15 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
     config.proxy.enabled = { docker: false }
   end
 
+  etcd_seed_cluster = Net::HTTP.get(URI.parse("https://discovery.etcd.io/new?size=%d" % (NODES.to_i + 1)))
```
Comment (pires) [inline Vagrantfile]: I don't like the fact we are now adding a new dependency, namely the availability of this service.

File: Vagrantfile [hunk 2]
```diff
@@ -75,7 +75,7 @@ DOCKERCFG = File.expand_path(ENV['DOCKERCFG'] || "~/.dockercfg")
 
 DOCKER_OPTIONS = ENV['DOCKER_OPTIONS'] || ''
 
-KUBERNETES_VERSION = ENV['KUBERNETES_VERSION'] || '1.6.2'
+KUBERNETES_VERSION = ENV['KUBERNETES_VERSION'] || '1.6.4'
```
Comment (pires) [inline Vagrantfile]: This shouldn't be part of this PR since it's not related to the upgrade to `etcd3`.

File: master.yaml [hunk 3]
```diff
@@ -36,13 +27,36 @@ coreos:
...
       enable: true
       command: start
+    - name: etcd3.service
+      command: start
+      content: |
+        [Unit]
+        Description=etcd
+        Documentation=https://github.com/coreos/etcd
+        [Service]
+        Environment='ETCD_IMAGE_TAG=latest'
```
Comment (pires) [inline master.yaml]: This should be the version officially supported by Kubernetes. Going with `latest` may mean in the feature that we run a new, incompatible version, e.g. `4.x`.

File: node.yaml [hunk 4]
```diff
@@ -33,13 +27,36 @@ coreos:
...
+        LimitNOFILE=40000
+        TimeoutStartSec=0
+        ExecStart=/usr/lib/coreos/etcd-wrapper --name __NAME__ \
+            --listen-client-urls http://0.0.0.0:2379,http://0.0.0.0:4001 \
+            --advertise-client-urls http://$public_ipv4:2379,http://$public_ipv4:4001 \
+            --listen-peer-urls http://$private_ipv4:2380,http://$private_ipv4:7001 \
+            --initial-advertise-peer-urls http://$private_ipv4:2380 \
+            --discovery __ETCD_SEED_CLUSTER__
+        [Install]
+        WantedBy=multi-user.target
```
Comment (pires) [inline node.yaml]: On worker nodes, `etcd` should be run in proxy mode.

File: Vagrantfile (hunk 2 above)
Comment (saschagrunert) [inline Vagrantfile]: Alright, I will revert it.

File: Vagrantfile (hunk 1 above)
Comment (saschagrunert) [inline Vagrantfile]: We could link the machines statically together via the parameters:
```
--initial-cluster infra0=http://10.0.1.10:2380,infra1=http://10.0.1.11:2380,infra2=http://10.0.1.12:2380 \
--initial-cluster-state new
```

File: master.yaml (hunk 3 above)
Comment (saschagrunert) [inline master.yaml]: Alright I will change it to 3.1.8

File: Vagrantfile (hunk 2 above)
Comment (pires) [inline Vagrantfile]: I actually changed this already in master so it should be fine.

File: Vagrantfile (hunk 1 above)
Comment (pires) [inline Vagrantfile]: Not really. If the worker nodes are only proxies, there's no need to define multiple etcd instances. This is a development tool that doesn't care about etcd high-availability (redundancy in this case is a waste of resources).

**Comment by saschagrunert** [comment]
Thank you for the review. I implemented your notes and it works now out of the box.

File: master.yaml [hunk 5]
```diff
@@ -36,13 +27,37 @@ coreos:
...
+        RestartSec=5s
+        LimitNOFILE=40000
+        TimeoutStartSec=0
+        ExecStart=/usr/lib/coreos/etcd-wrapper --name __NAME__ \
+            --listen-client-urls http://0.0.0.0:2379,http://0.0.0.0:4001 \
+            --advertise-client-urls http://$public_ipv4:2379,http://$public_ipv4:4001 \
+            --listen-peer-urls http://$private_ipv4:2380,http://$private_ipv4:7001 \
+            --initial-advertise-peer-urls http://$private_ipv4:2380 \
+            --initial-cluster __ETCD_SEED_CLUSTER__ \
+            --initial-cluster-state new
```
Comment (pires) [inline master.yaml]: Remove this.
